Jenkinsfile

```diff
@@ -1,150 +1,100 @@
           }
-        }
-        sh 'make install'
-        sh 'make code-coverage-before-build'
-        sh 'make tests'
-        sh 'make code-coverage-after-build'
 
-        withCredentials([string(credentialsId: 'psammead-chromatic-app-code', variable: 'CHROMATIC_APP_CODE')]) {
-          sh 'npm run test:chromatic'
-        }
-      }
-      post {
-        always {
-          script {
-            stageName = env.STAGE_NAME
+          stage ('Development Tests') {
+            parallel (
+              'App Tests & Code Coverage': {
+                sh 'make code-coverage-before-build'
+                sh 'make tests'
+                sh 'make code-coverage-after-build'
+              },
+              'ChromaticQA Tests': {
+                withCredentials([string(credentialsId: 'psammead-chromatic-app-code', variable: 'CHROMATIC_APP_CODE')]) {
+                  sh 'npm run test:chromatic'
+                }
+              }, failFast: true
+            )
           }
-        }
-      }
-    }
-    stage ('Deploy Storybook & Publish to NPM') {
-      when {
-        expression { env.BRANCH_NAME == 'latest' }
-      }
-      agent {
-        docker {
-          image "${nodeImage}"
-          label nodeName
-          args '-u root -v /etc/pki:/certs'
-        }
-      }
-      steps {
-        sh 'ls' // quick debug step
-        sh 'make storybook'
-        sh 'rm published.txt || true'
-        withCredentials([string(credentialsId: 'npm_bbc-online_read_write', variable: 'NPM_TOKEN')]) {
-          sh 'make publish'
-        }
-        stash name: 'psammead-publishes', includes: 'published.txt', allowEmpty: true
-        sh 'rm published.txt || true'
-      }
-      post {
-        always {
-          script {
-            stageName = env.STAGE_NAME
+
+          if (env.BRANCH_NAME == 'latest') {
+            stage ('Deploy Storybook & Publish to NPM') {
+              parallel (
+                'Deploy Storybook': {
+                  sh 'npm run deploy-storybook'
+                },
+                'Publish to NPM': {
+                    withCredentials([string(credentialsId: 'npm_bbc-online_read_write', variable: 'NPM_TOKEN')]) {
+                      sh 'make setup-npm'
+                      sh 'npm run publish'
+                    }
+                }
+              )
+            }
+
+            stage ('Bump Dependants') {
+              sh 'npm run updateDependants'
+            }
           }
-        }
-      }
-    }
-    stage ('Bump Dependants') {
-      when {
-        expression { env.BRANCH_NAME == 'latest' }
-      }
-      agent {
-        docker {
-          image "${nodeImage}"
-          label nodeName
-          args '-u root -v /etc/pki:/certs'
-        }
-      }
-      steps {
-        unstash 'psammead-publishes'
-        sh 'make bump-dependants'
-      }
-      post {
-        always {
-          script {
-            stageName = env.STAGE_NAME
+        } catch (Throwable e) {
```

Comment:
Not familiar with Groovy exception handling but assuming it works like Java won't this result in a double Slack notification being sent as the `finally` block will also run?

Also is the `Throwable` here redundant?

Reply:
Was having having issues where the catch was not actually catching pipeline errors until I used `Throwable e`

It is true the the `finally` block will always run, this is why the default in the switch statement does not send a slack notification it simply echos the build status.

The switch case in the finally block also ensure we only send a slack notification if the build status is `SUCCESS` or `UNSTABLE` failed states will be caught in the `catch`

Reply:
OK - though I still don't quite get it, if the build is `UNSTABLE` won't you get both a `Failed` and an `Unstable` Slack notification?

Reply:
hmmmmm I am not actually 100% sure. I need to see what triggers the `unstable` state.

if that is the case we can check the build status in the catch. And only have the finally block post to slack if the build status is a `success`?

Reply:
@jamesdonoh https://ci.news.tools.bbc.co.uk/blue/organizations/jenkins/psammead/detail/pipeline-improvemnets/45/pipeline/28

![image](https://user-images.githubusercontent.com/14273378/62962761-7b9d7800-bdf7-11e9-9744-285f0ce1acfd.png)


I forced this pipeline into an `UNSTABLE` state. This was not caught by the catch.

The notification was sent from the switch statement in the finally `block`

Reply:
Perhaps the success/unstable slack notifications could be moved into the main 'try' block, or after the finally block? That way the failure execution flow shouldn't go anywhere near them

Reply:
I dont think we can do this, as an unstable will be caught by the try catch no matter what

Reply:
if unstables get caught by the try/catch block, then why don't we switch on `currentBuild.result` and slack failures/unstables from there?